src/java/org/apache/cassandra/io/sstable/format/SSTableReader.java

```diff
@@ -1479,6 +1479,31 @@ public List<PartitionPositionBounds> getPositionsForRanges(Collection<Range<Toke
         return positions;
     }
 
+    public long onDiskSizeForRanges(Collection<Range<Token>> ranges)
+    {
+        long total = 0;
+        if (!compression)
+        {
+            for (var position : getPositionsForRanges(ranges))
+                total += position.upperPosition - position.lowerPosition;
+        }
+        else
+        {
+            final CompressionMetadata compressionMetadata = getCompressionMetadata();
+            long lastEnd = 0;
+            for (var position : getPositionsForRanges(ranges))
+            {
+                long upperChunkEnd = compressionMetadata.chunkFor(position.upperPosition).chunkEnd();
+                long lowerChunkStart = compressionMetadata.chunkFor(position.lowerPosition).offset;
+                if (lowerChunkStart < lastEnd)  // if regions include the same chunk, count it only once
+                    lowerChunkStart = lastEnd;
+                total += upperChunkEnd - lowerChunkStart;
+                lastEnd = upperChunkEnd;
```

Comment:
Looks like `Range.normalize` ensures the ranges don't overlap. It seems like this code is being defensive in case that implementation changes. In that case, should we also add something like the following:

```java
if (upperChunkEnd <= lastEnd)
    continue;
```

Reply:
Perhaps there is something about compression metadata that I'm not familiar with here?

Reply:
If we have two sections immediately (or close after) one another, they may fall in the same compression chunk. In that case the lower bound is its start and the upper its end. Without this adjustment the chunk would be counted twice.